### PR TITLE
HelpersTask465_generalize_increase_chapter

### PR DESCRIPTION
**Pull request by allenmatt10**
Related to #465 

Generalized `increase_chapter()` in `helpers/hmarkdown.py` to decrease header levels.

Modified and added new test cases in `helpers/test/test_hmarkdown.py`.

### CODE REVIEW

**Comment by sonniki** [comment]
@samarth9008 could you please do the first review?

File: helpers/hmarkdown.py [hunk 1]
```diff
@@ -542,23 +542,36 @@ def format_headers(in_file_name: str, out_file_name: str, max_lev: int) -> None:
 # TODO(gp): Add tests.
 # TODO(gp): Generalize this to also decrease the header level
 # TODO(gp): -> modify_header_level
```
Comment (samarth9008) [inline helpers/hmarkdown.py:544]: We should remove the TODO if they are resolved to avoid confusion for future.

File: helpers/hmarkdown.py [hunk 2]
```diff
@@ -542,23 +542,36 @@ def format_headers(in_file_name: str, out_file_name: str, max_lev: int) -> None:
...
-def increase_chapter(in_file_name: str, out_file_name: str) -> None:
+def modify_header_level(
+    in_file_name: str, out_file_name: str, signal: int
+) -> None:
     """
-    Increase the level of chapters by one for text in stdin.
+    Increase or decrease the level of chapters by one for text in stdin.
```
Comment (samarth9008) [inline helpers/hmarkdown.py:549]: chapters -> headings

File: helpers/hmarkdown.py [hunk 3]
```diff
@@ -542,23 +542,36 @@ def format_headers(in_file_name: str, out_file_name: str, max_lev: int) -> None:
...
 
-    :param in_file_name: The name of the input file to read
+    The function reads the input file, adjusts the chapter header levels
+    by one step based on the provided signal, and writes the modified
+    content to the output file.
```
Comment (samarth9008) [inline helpers/hmarkdown.py:553]: no need for this

File: helpers/hmarkdown.py [hunk 4]
```diff
@@ -542,23 +542,36 @@ def format_headers(in_file_name: str, out_file_name: str, max_lev: int) -> None:
 # TODO(gp): Add tests.
 # TODO(gp): Generalize this to also decrease the header level
 # TODO(gp): -> modify_header_level
-def increase_chapter(in_file_name: str, out_file_name: str) -> None:
+def modify_header_level(
+    in_file_name: str, out_file_name: str, signal: int
```
Comment (samarth9008) [inline helpers/hmarkdown.py:546]: Instead of signal better to use a bool variable such as `increase: bool = True` (default behavior), since that was how the code behaved previously. If increase is false, we decrease. This way we won't need to change the code at many places.

File: helpers/hmarkdown.py [hunk 5]
```diff
@@ -542,23 +542,36 @@ def format_headers(in_file_name: str, out_file_name: str, max_lev: int) -> None:
...
+        (-1) of the header level.
     """
     txt = hparser.read_file(in_file_name)
     #
     txt_tmp = []
     for line in txt:
         # TODO(gp): Use the iterator.
         line = line.rstrip(r"\n")
-        for i in range(1, 5):
+        for i in range(1, 7):
```
Comment (samarth9008) [inline helpers/hmarkdown.py:567]: Why did we change it from 5 to 7? I would keep `range(1,6)` bcz max `#` we use are 5. We should also handle and test the edge case where we can't decrease if `#` is 1 or increase its 5

**Review by samarth9008 — COMMENTED** [review]
Mid-level review as the code requires some arch changes.

Modify/improve/add news tests if needed.

File: helpers/hmarkdown.py (hunk 5 above)
Comment (allenmatt10) [reply helpers/hmarkdown.py:567]: I changed it to 7 because Markdown headings range from 1 to 6. I will update the code and make the necessary changes.

Comment (samarth9008) [reply helpers/hmarkdown.py:567]: lets keep the old range of `range(1,5)`

Comment (allenmatt10) [reply helpers/hmarkdown.py:567]: In that case, for test cases decreasing headers with size 5 will stay the same. Keeping range(1,6) solves all possible cases. Shall I keep the old range or update it to `range(1,6)`?

**Comment by allenmatt10** [comment]
I've addressed the fixes required. Also, I have set the range in the function which satisfies the following edge cases:

- Increase for headings with 5 `#` to stay the same.
- Decrease for headings with 1 `#` to stay the same.
- Decrease for headings with 5 `#` will reduce by 1.

Let me know if any changes are to be made.

**Review by sonniki — CHANGES_REQUESTED** [review]

File: helpers/hmarkdown.py [hunk 6]
```diff
@@ -539,26 +539,35 @@ def format_headers(in_file_name: str, out_file_name: str, max_lev: int) -> None:
     hparser.write_file(txt_tmp, out_file_name)
 
 
-# TODO(gp): Add tests.
-# TODO(gp): Generalize this to also decrease the header level
-# TODO(gp): -> modify_header_level
-def increase_chapter(in_file_name: str, out_file_name: str) -> None:
+def modify_header_level(
+    in_file_name: str, out_file_name: str, increase: bool = True
```
[the remark this answers is not in the document]
Comment (samarth9008) [reply helpers/hmarkdown.py:543]: > IMO it's less clear to have increase=False indicating that the level should be decreased, since "not increasing" != "decreasing".

Better, sg

File: helpers/hmarkdown.py [hunk 7]
```diff
@@ -539,26 +539,35 @@ def format_headers(in_file_name: str, out_file_name: str, max_lev: int) -> None:
...
     """
     txt = hparser.read_file(in_file_name)
     #
     txt_tmp = []
+    increase_level = 1
     for line in txt:
         # TODO(gp): Use the iterator.
         line = line.rstrip(r"\n")
-        for i in range(1, 5):
+        for i in range(1, 6):
```
[the remark this answers is not in the document]
Comment (sonniki) [reply helpers/hmarkdown.py:561]: Feel free to propose a solution based on your own ideas of what would be best.

[the remark this answers is not in the document]
Comment (allenmatt10) [reply helpers/hmarkdown.py:561]: @sonniki 
Proposed solutions:

- We can modify `is_markdown_line_separator()` to exclude headings with 5`#`. Currently,  the test cases for this function do not test for headings with 5 `#`. Also, It's function calls would not have any serious impact if we follow this solution, so this change should be safe.
- We can use the iterator which handles all possible edge cases, as implemented in the previous commit.

[the remark this answers is not in the document]
Comment (sonniki) [reply helpers/hmarkdown.py:561]: Go for num.1

File: dev_scripts_helpers/documentation/transform_notes.py [hunk 8]
```diff
@@ -74,7 +74,7 @@ def _main(parser: argparse.ArgumentParser) -> None:
     elif cmd == "format":
         hmarkdo.format_headers(in_file_name, out_file_name, max_lev)
     elif cmd == "increase":
-        hmarkdo.increase_chapter(in_file_name, out_file_name)
+        hmarkdo.modify_header_level(in_file_name, out_file_name, cmd)
```
Comment (sonniki) [inline dev_scripts_helpers/documentation/transform_notes.py:77]: Funny coincidence that here `cmd` is also "increase" but I would be careful -- if this code here is later refactored and `cmd` stops being exactly "increase", we'll have a problem. Just indicating `mode="increase"` is safer.

File: helpers/hmarkdown.py [hunk 9]
```diff
@@ -539,27 +539,38 @@ def format_headers(in_file_name: str, out_file_name: str, max_lev: int) -> None:
...
     """
-    Increase the level of chapters by one for text in stdin.
+    Increase or decrease the level of headings by one for text in stdin.
 
-    :param in_file_name: The name of the input file to read
-    :param out_file_name: The name of the output file to write the
+    :param in_file_name: the name of the input file to read
+    :param out_file_name: the name of the output file to write the
         modified text to
+    :param mode: a string indicating the increase or decrease of the
```
Comment (sonniki) [inline helpers/hmarkdown.py:549]: We usually avoid referring to type of vars in docstrings (e.g. "string"), since it can be already seen from the typing hints.

**Review by sonniki — CHANGES_REQUESTED** [review]
A couple of small things left

**Review by samarth9008 — COMMENTED** [review]
LGTM
